src/addrman.h

```diff
@@ -534,137 +559,145 @@ friend class CAddrManTest;
 
     //! Add a single address.
-    bool Add(const CAddress &addr, const CNetAddr& source, int64_t nTimePenalty = 0)
+    bool Add(const CAddress &addr, const CNetAddr& source, int64_t nTimePenalty = 0) EXCLUSIVE_LOCKS_REQUIRED(!m_addrman_mutex)
     {
-        LOCK(cs);
+        AssertLockNotHeld(m_addrman_mutex);
+        LOCK(m_addrman_mutex);
         bool fRet = false;
-        Check();
+        CheckNonLockHelper();
         fRet |= Add_(addr, source, nTimePenalty);
-        Check();
+        CheckNonLockHelper();
         if (fRet) {
             LogPrint(BCLog::ADDRMAN, "Added %s from %s: %i tried, %i new\n", addr.ToStringIPPort(), source.ToString(), nTried, nNew);
         }
         return fRet;
     }
 
     //! Add multiple addresses.
-    bool Add(const std::vector<CAddress> &vAddr, const CNetAddr& source, int64_t nTimePenalty = 0)
+    bool Add(const std::vector<CAddress> &vAddr, const CNetAddr& source, int64_t nTimePenalty = 0) EXCLUSIVE_LOCKS_REQUIRED(!m_addrman_mutex)
     {
-        LOCK(cs);
+        AssertLockNotHeld(m_addrman_mutex);
+        LOCK(m_addrman_mutex);
         int nAdd = 0;
-        Check();
+        CheckNonLockHelper();
         for (std::vector<CAddress>::const_iterator it = vAddr.begin(); it != vAddr.end(); it++)
             nAdd += Add_(*it, source, nTimePenalty) ? 1 : 0;
-        Check();
+        CheckNonLockHelper();
         if (nAdd) {
             LogPrint(BCLog::ADDRMAN, "Added %i addresses from %s: %i tried, %i new\n", nAdd, source.ToString(), nTried, nNew);
         }
         return nAdd > 0;
     }
 
     //! Mark an entry as accessible.
-    void Good(const CService &addr, bool test_before_evict = true, int64_t nTime = GetAdjustedTime())
+    void Good(const CService &addr, bool test_before_evict = true, int64_t nTime = GetAdjustedTime()) EXCLUSIVE_LOCKS_REQUIRED(!m_addrman_mutex)
     {
-        LOCK(cs);
-        Check();
+        AssertLockNotHeld(m_addrman_mutex);
+        LOCK(m_addrman_mutex);
+        CheckNonLockHelper();
         Good_(addr, test_before_evict, nTime);
-        Check();
+        CheckNonLockHelper();
     }
 
     //! Mark an entry as connection attempted to.
-    void Attempt(const CService &addr, bool fCountFailure, int64_t nTime = GetAdjustedTime())
+    void Attempt(const CService &addr, bool fCountFailure, int64_t nTime = GetAdjustedTime()) EXCLUSIVE_LOCKS_REQUIRED(!m_addrman_mutex)
     {
-        LOCK(cs);
-        Check();
+        AssertLockNotHeld(m_addrman_mutex);
+        LOCK(m_addrman_mutex);
+        CheckNonLockHelper();
         Attempt_(addr, fCountFailure, nTime);
-        Check();
+        CheckNonLockHelper();
     }
 
     //! See if any to-be-evicted tried table entries have been tested and if so resolve the collisions.
-    void ResolveCollisions()
+    void ResolveCollisions() EXCLUSIVE_LOCKS_REQUIRED(!m_addrman_mutex)
     {
-        LOCK(cs);
-        Check();
+        AssertLockNotHeld(m_addrman_mutex);
+        LOCK(m_addrman_mutex);
```

Comment:
> Following that, should we add `AssertLockNotHeld()` before every `LOCK()` in the source code?

Only for `Mutex` instances, not for `RecursiveMutex` ones.

> What about adding `AssertLockNotHeld()` inside `LOCK()`?

Not now because the `RecursiveMutex` instance could be an argument of `LOCK()`.

Reply:
> What about checking, inside `LOCK()`, whether the mutex being locked is instance of `Mutex` (and not `RecursiveMutex`) and only asserting that we don't own it in that case

It is not thread safe, for sure.

Reply:
https://softwareengineering.stackexchange.com/questions/330769/why-cant-i-check-whether-a-mutex-is-locked

Reply:
@vasild Mind submitting a PR that it could be reviewed before this one?

Reply:
Currently, it is assumed the following patterns:
- for public interface methods:
```
# foo.h
class Foo
{
public: void foo() EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
}
# foo.cpp
void Foo::foo()
{
    AssertLockNotHeld(m_mutex);
    LOCK(m_mutex);
    // do magic
}
```

- for internal methods that are called from critical sections:
```
# foo.h
class Foo
{
pivate: void foo_cs() EXCLUSIVE_LOCKS_REQUIRED(m_mutex);
}
# foo.cpp
void Foo::foo_cs()
{
    AssertLockHeld(m_mutex);
    // do magic
}
```

The common places are thread safety annotations and lock assertions. It seems placing `AssertLockNotHeld()` into internals of the `LOCK()` breaks uniformity, no?

Reply:
> Mind submitting a PR that it could be reviewed before this one?

Yes, I think that deserves a separate PR. I have to consider this a little bit before opening a PR. Maybe it would be possible to optimize it - the patch above would lock `lockdata.dd_mutex` and dig into the stack of locks for the current thread 2 times - once from the newly added `AssertLockNotHeld()` and a second time from `push_lock()`. It should be possible to do everything in `push_lock()` and avoid adding the call to `AssertLockNotHeld()`.

> ...breaks uniformity, no?

I don't see it that way - `LOCK()` already implies that we don't own the lock.

Reply:
Here is a PR that would detect a double lock. With the check from inside `LOCK()` it makes unnecessary to add `AssertLockNotHeld()` calls before each `LOCK()`: https://github.com/bitcoin/bitcoin/pull/19337